### ISSUE TITLE
eslint semi, omitLastInOneLineBlock: true for JSX statements

### ISSUE THREAD

**Issue by CostachescuCristinel** [eslint/eslint#12536]
**What rule do you want to change?**
I would like to ask a change of the default behavior of the `semi` rule for JSX statements, when the option `"always"` is used together with the `"omitLastInOneLineBlock": true` option.
For JSX statements used in a function's `return` statement, the rule will require a semicolon after the closing JSX tag (or the self-closing JSX tag). This is JS-correct as a return statement should require a semicolon. However, a JSX closing tag (or self-closing tag) is explicitly implying the end of the statement. As such, it should not require semicolon.

**Does this change cause the rule to produce more or fewer warnings?**
The change will produce **less** warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
A new default behavior should be in effect.
Please note however that, the way it is meant to behave after the change, will not break the current default behavior.

**Please provide some example code that this change will affect:**

```js
function example(){
    return <SomeJSXComponent>
        ...
    </SomeJSXComponent>    // <--- here eslint rule semi with "omitLastInOneLineBlock": true still requires a semicolon
}
function example_selfClosingTag(){
    return <SomeJSXComponent />    // <--- here eslint rule semi with "omitLastInOneLineBlock": true still requires a semicolon
}
```

**What does the rule currently do for this code?**
The current rule requires a semicolon after the closing JSX tag.

**What will the rule do after it's changed?**
The rule will follow the `"omitLastInOneLineBlock": true` option behavior for JSX tags the same way it does for non-JSX code.

**Are you willing to submit a pull request to implement this change?**
I am not in the position of managing a pull request.

**Comment by kaicataldo** [comment]
Hi, can you explain why `omitLastInOneLineBlock` should cover this case? The example given isn’t a block and is multiline.

**Comment by CostachescuCristinel** [comment]
I think it should cover this case because ultimately, after the JSX is handled, it is converted into a simple function call. Considering a React use-case, for example, the JSX from above is converted into:
```
    return React.createElement(SomeJSXComponent, ...properties )
```
This conversion, although not visually noticeable when looking at the source code itself, will in fact take place right before the JS parsing step, where the JSX is converted into actual JavaScript code.
In a more general use case than React, JSX is converted into an element that is immediately understandable by the underlying language; it could be an atomic value (a number, a string, ect), a simple data type, but the most common convertion is into a simple function call - all of them translate into code that does not spawn accross multiple separate statements (so, you will never get JSX converted into `var i=0; return i;`) - thus, one line block.

Hopefully my explanation gives a valid reason, but feel free to point out if I am mistaken.

**Comment by kaicataldo** [comment]
I’m sorry, I’m unfortunately not following. I still don’t quite understand what this has to do with the `omitLastInOneLineBlock` option, since the change doesn’t related to either blocks or single line statements.

I’m also not understanding why the compiled output should determine the style for the uncompiled source. If this is something you’d like, I think it makes sense to create your own version of the rule using [`eslint-rule-composer`](https://github.com/not-an-aardvark/eslint-rule-composer).

**Comment by CostachescuCristinel** [comment]
I understand, thank you for your time!
